File: frontend/src/features/actions/types/action.ts
export type ActionPolicy = "auto" | "needs_human";

export interface ActionRecipeStep {
  id: string;
  /** Tool instance id (enabled binding). Legacy: type id when single instance. */
  tool: string;
  /** Command within the tool, e.g. run_structured / search. */
  command: string;
  /** Step input object with {{template}} expressions. */
  input: Record<string, unknown>;
  /** Optional guard expression; step runs only when true. */
  when?: string;
}

export interface ActionDef {
  id: string;
  name: string;
  description: string;
  version: string;
  policy: ActionPolicy;
  inputSchema: Record<string, unknown>;
  outputSchema: Record<string, unknown>;
  recipe: ActionRecipeStep[];
  createdAt: string;
  updatedAt: string;
}

export type ActionInput = {
  id: string;
  name: string;
  description?: string;
  version?: string;
  policy?: ActionPolicy;
  inputSchema?: Record<string, unknown>;
  outputSchema?: Record<string, unknown>;
  recipe?: ActionRecipeStep[];
};

export type ActionMutationResult =
  | { ok: true; action: ActionDef }
  | { ok: false; error: string };

File: backend/domain/action.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

ActionPolicy = Literal["auto", "needs_human"]


class ActionRecipeStep(BaseModel):
    id: str = ""
    tool: str
    command: str
    input: dict[str, Any] = Field(default_factory=dict)
    when: str = ""


class ActionDefinition(BaseModel):
    id: str
    name: str = ""
    description: str = ""
    version: str = "0.1.0"
    policy: ActionPolicy = "auto"
    on_failure: str = "fail"
    retry: dict[str, Any] = Field(default_factory=dict)
    timeout_ms: int | None = None
    input_schema: dict[str, Any] = Field(default_factory=dict, alias="inputSchema")
    output_schema: dict[str, Any] = Field(default_factory=dict, alias="outputSchema")
    recipe: list[ActionRecipeStep] = Field(default_factory=list)

    model_config = {"populate_by_name": True}
